reading_comprehension/qanet.py

```diff
@@ -0,0 +1,256 @@
+
+        encoded_question = self._dropout(self._phrase_layer(projected_embedded_question, question_mask))
+        encoded_passage = self._dropout(self._phrase_layer(projected_embedded_passage, passage_mask))
+
+        # Shape: (batch_size, passage_length, question_length)
+        passage_question_similarity = self._matrix_attention(encoded_passage, encoded_question)
+        # Shape: (batch_size, passage_length, question_length)
+        passage_question_attention = masked_softmax(passage_question_similarity, question_mask)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        passage_question_vectors = util.weighted_sum(encoded_question, passage_question_attention)
+
+        # Shape: (batch_size, question_length, passage_length)
+        question_passage_attention = masked_softmax(passage_question_similarity.transpose(1, 2), passage_mask)
+        # Shape: (batch_size, passage_length, passage_length)
+        attention_over_attention = torch.bmm(passage_question_attention, question_passage_attention)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        passage_passage_vectors = util.weighted_sum(encoded_passage, attention_over_attention)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4)
+        merged_passage_attention_vectors = self._dropout(
+                torch.cat([encoded_passage, passage_question_vectors,
+                           encoded_passage * passage_question_vectors,
+                           encoded_passage * passage_passage_vectors],
+                          dim=-1)
+                )
+
+        modeled_passage_list = [self._modeling_proj_layer(merged_passage_attention_vectors)]
+
+        for _ in range(3):
+            modeled_passage = self._dropout(self._modeling_layer(modeled_passage_list[-1], passage_mask))
+            modeled_passage_list.append(modeled_passage)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4 + modeling_dim))
+        span_start_input = torch.cat([modeled_passage_list[-3], modeled_passage_list[-2]], dim=-1)
+        # Shape: (batch_size, passage_length)
+        span_start_logits = self._span_start_predictor(span_start_input).squeeze(-1)
+        # Shape: (batch_size, passage_length)
+        span_start_probs = masked_softmax(span_start_logits, passage_mask)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4 + span_end_encoding_dim)
+        span_end_input = torch.cat([modeled_passage_list[-3], modeled_passage_list[-1]], dim=-1)
+        span_end_logits = self._span_end_predictor(span_end_input).squeeze(-1)
+        span_end_probs = masked_softmax(span_end_logits, passage_mask)
+        span_start_logits = util.replace_masked_values(span_start_logits, passage_mask, -1e7)
+        span_end_logits = util.replace_masked_values(span_end_logits, passage_mask, -1e7)
+        best_span = BidirectionalAttentionFlow.get_best_span(span_start_logits, span_end_logits)
```

Comment:
Looks like we should pull `get_best_span` out into a `reading_comprehension.utils` module.  But that can be done later.